Add some unit tests

fractal161 opened the pull request:
features a handful of typo fixes, a new migration, and an implemented testing environment with docker. ah, and actual test cases too.

right now the testing suite is incomplete in that it only tests correct queries. eventually i plan to add those in, but that seems like much effort so i wanna check that everything looks good first.

fractal161 committed:
fix arguments for resolvers

Several resolver methods were bugged because prisma expects a particular
order of function arguments (i.e. parent, args, context), and the args
parameter was not included for resolvers that didn't use any arguments.

fractal161 committed:
add test db as docker container

This summary was added retroactively; as such, I'm not completely sure
about the justification for each change.

In essence, the testing strategy spins up a brand new postgres database,
runs migrations, then the actual test commands, and finally destroys the
test database. If my memory is correct, we map the test container's port 5432 to 5433 to distinguish between the URLs of the test and dev database.

I also changed the main server code to listen to all interfaces, so that
it could connect to the test database in docker. This should probably be
changed eventually.

fractal161 committed:
refactor test structure

This commit separates out tests for each query type into their own
files, which are then run sequentially in the main test. The other
primary addition is a file to store global variables, which will be used
to validate certain information (e.g. objects ids).

fractal161 committed:
add creators for player and event

Adds affirmative tests for createPlayer, createPlayercomplete, and
createEvent.

Couple of other fixes/changes too. First, the edition in the Event type
was made nullable to be consistent with the prisma schema. Second, typo
in createPlayerComplete was fixed (replaced "counrty" with "country").

fractal161 committed:
complete unit tests for creators

Adds affirmative tests for createMatchByEventName, createMatchByEventId,
createGame, createResultByPlayerName, and createResultByPlayerId.

Schema/api changes: rom parameter in createMatchByEventName made nullable to
match prisma schema, styles parameter in Result type added to match
prisma schema, createResultByPlayerId passes correct parameters to
prisma.

fractal161 committed:
add tests for getters

Add affirmative tests for getOrganizationByName, getOrganizationById,
getAllOrganizations, getEventByName, getEventById, getEventsByEdition,
getAllMatches, getMatchesByPlayer, getPlayerByName, getPlayerById, and
getAllPlayers.

api changes: rename the "name" parameter in getPlayerByName to
"eloName", since this is what's actually being queried.

fractal161 committed:
make getEventByName not return an array

Since events should have unique names (which is indeed enforced in the
prisma schema), it doesn't make sense to return an array. Thus, this
commit changes getEventByName to return a single (possibly null) Event.

fractal161 committed:
add tests for updaters

Adds affirmative tests for updatePlayerName, addPlaystyle, and
removePlaystyle.

fractal161 committed:
add tests for non-elo resolvers

Tests that the non-elo resolvers for Events, Matches, Games, Organizations,
Players, and Results all work as intended. Not covered are resolvers for
EloSnapshots and elo information for Events and Players (but should be
added eventually).

fractal161 committed:
rename playstyle field in results

Slight API change: all other types (e.g. Player) refer to this
information as "playstyle," so "style" was renamed to be consistent.

The corresponding migration was manually edited to rename the column, since the
default behavior is to delete the existing column/replace it with a new
one, which wasn't necessary here.

professor-l commented on docker-compose.yml:
```diff
@@ -11,6 +11,16 @@ services:
+      POSTGRES_PASSWORD: "${POSTGRES_PASSWORD}"
+      POSTGRES_DB: "ctdb"
+    ports:
+      - "5433:5432"
```
Any particular reason we expose 5433 instead of 5432? I don't think it's a problem but I am curious.

fractal161 replied:
I want to say that this was so that the dev and test databases wouldn't be exposing the same port? Though I don't even know if this is how ports work so could be totally off base. Will make a note in the modified commit msgs

professor-l commented on package.json:
```diff
@@ -7,7 +7,7 @@
     "dev": "dotenv -e .env.development -- ts-node-dev --exit-child src/index.ts",
     "start": "dotenv -e .env.development -- ts-node src/index.ts",
-    "test": "dotenv -e .env.test -- prisma migrate reset --force --skip-seed && dotenv -e .env.test -- jest --config jest.config.ts ./test --detectOpenHandles",
+    "test": "docker-compose up -d test_db && sleep 1 && dotenv -e .env.test -- prisma migrate reset --force --skip-seed && dotenv -e .env.test -- jest --config jest.config.ts ./test --detectOpenHandles && docker-compose stop test_db && docker-compose rm -f test_db",
```
Might be worth separating this out to its own `scripts/test.sh` file, just for legibility's sake. Would also make debugging easier as with `&&` each command is dependent on the previous command exiting with status code 0. A separate script would let us error out more verbosely.

We could also change the `sleep 1` into a looping liveness probe so we can just run the tests whenever the db is ready.

professor-l commented on test/creators.ts:
It may be worth separating out the `test_cases` to a separate object, maybe some kind of map of `description -> testData` or something, and then looping over that object to actually run the tests. This would:

- Improve legibility of the existing test cases
- Ease the burden of adding new test cases

However this is a significant refactor so it would make  sense to me to tackle it in a separate PR.

fractal161 replied:
Probably gonna leave this for a later pr. Each test's structure is consistent, so I don't think this will ultimately be too annoying.

professor-l commented on test/testState.ts:
This is another piece that may be worth refactoring in a future PR - we talked in VC about the possibility of using a script and `faker.js` to pre-generate a static file of test data. This would:

- let us test on larger datasets, making the test behavior closer to what we'd see in production
- avoid the randomness element of dynamically generating test data on each run
- allow us to tweak or add to the data manually if and when we need to

professor-l commented on prisma/migrations/20230620012446_rename_playstyle_field_in_results/migration.sql:
Worth noting in the commit message that this was a custom migration because by default the one Prisma generated deleted and re-created the field with its new name, which would lose data.

professor-l reviewed (approved):
This looks fantastic! Other than more detail in some of the commit messages I'd say this is ready to be merged - we can absolutely look at future refactors for some of the clunkier bits but nothing is blocking, so I'll leave an approval and let you touch up some of the commit messages whenever you have time.

Thank you so much for all this work!!

fractal161 committed:
Move test script to separate file

This commit also removes unnecessary print statements in the tests.

I originally needed a `sleep 1` command to wait for the container to
become ready, but I tried removing it and I haven't had any issues, so I
will leave this out for now and hope that the problems don't resurface.

fractal161 commented:
added some more detail to all the existing commits and split out the test script to its own file. will go ahead and merge.

fractal161 commented:
lmaooo i forgot to stage one line (package.json calls `./test.sh` instead of `scripts/test.sh`), will fix that
